File: EmotionEngine/src/main/java/co/q64/emotion/ast/ASTInstruction.java
package co.q64.emotion.ast;

import javax.annotation.Nullable;
import javax.inject.Inject;

import com.google.auto.factory.AutoFactory;

import co.q64.emotion.lang.Instruction;
import co.q64.emotion.lang.Program;

@AutoFactory
public class ASTInstruction implements ASTNode {
	private Program program;
	private Instruction insn;

	protected @Inject ASTInstruction(@Nullable Program program, Instruction insn) {
		this.program = program;
		this.insn = insn;
	}

	@Override
	public ASTBackpropagation enter() {
		if (program.shouldContinueExecution()) {
			try {
				insn.execute(program.getStack());
			} catch (Exception e) {
				program.crash(e.getClass().getSimpleName() + ": " + e.getMessage() + " [Instruction: " + insn.getInstruction() + "]");
			}
		}
		return ASTBackpropagation.NONE;
	}

	@Override
	public String toString() {
		return insn.getInstruction();
	}
}

File: emotion-core/src/main/kotlin/co/q64/emotion/core/ast/ASTInstruction.kt
package co.q64.emotion.core.ast

import co.q64.emotion.core.lang.Instruction
import co.q64.emotion.core.lang.Program

class ASTInstruction(
    private val program: Program,
    private val instruction: Instruction
) : ASTNode {
    override fun enter(): ASTResult {
        if (!program.terminated()) {
            try {
                instruction.execute(program)
            } catch (e: Exception) {
                program.crash("${e::class.java.simpleName}: ${e.message} [Instruction: $instruction]")
            }
        }
        return ASTResult.None
    }

    override fun toString() = "$instruction"
}